simulator/func_sim/func_sim.h

```diff
@@ -35,6 +35,7 @@ class FuncSim : public Simulator
         uint64 sequence_id = 0;
         std::shared_ptr<FuncMemory> mem;
         InstrMemoryCached<FuncInstr> imem;
+        std::weak_ptr<Kernel> kernel;
```

Comment:
Is this a necessity to have at least one `shared_ptr`, or is it just for a good code style? I've done a few tests with two `weak_ptr`'s, and it seems to work as expected

Reply:
> I've done a few tests with two weak_ptr's, and it seems to work as expected

And what releases memory?

Reply:
> And what releases memory?

The owner of initial `shared_ptr`'s, in our case it's `main()`.

Reply:
Then it is not a shared ownership, and you may still use `std::unique_ptr` here or call `delete` explicitly. The motivation behind `std::shared_ptr` is that you have not to care about memory freeing, it would be performed automatically by the last agent holding a resource.